### ISSUE TITLE
Request Project Permissions Failure

### ISSUE THREAD

**Issue by austintruchan** [matrix-msu/Kora3#348]
I created a new account on kora3.matrix.msu.edu and requested permissions to a project. However after doing so, the loading icon appeared and animated for ~5mins, ultimately providing this: 
<img width="425" alt="screen shot 2018-10-18 at 2 34 21 pm" src="https://user-images.githubusercontent.com/5408529/47176441-a015ea00-d2e3-11e8-8cd6-5868c7939749.png">

The email however did go through to the admin of the project I was requesting permissions to.

**Comment by Dev34** [comment]
Firstly, I was unable to reproduce this bug on my dev so I can't do much.
So the request emails are at least getting through to the project's admin group. Then it is failing with status code 504 Gateway Timeout. This implies that the mail server is timing out one of the requests - not sure why though.

One thing to consider is that when requesting permission to a project, it first sends the request email to the project's admin group, but then also sends the request email to all Kora3 admins on the installation. It seems like the timeout is happening when trying to send an email to one of the Kora3 admins who are not part of the project's admin group. If I could debug this I'd look to see exactly when the 504 is occurring, and if it is related to a particular email, or maybe the volume of mail server requests. I can't debug this though, so I'm going to unassign myself.

**Comment by fritosxii** [comment]
Rainer, I'm assigning this to you, but I'll try to help replicate as well

**Comment by fritosxii** [comment]
If it is a timeout issue, it's likely the Laravel mail call (which is not a quick operation) is timing out getting through all the admins. In the ProjectController request function, see if you can modify it to send one email to multiple people, rather than one email per admin. Ideally, we would want it to email to the root user (with id=1) and then BCC the other admins

**Comment by Dev34** [comment]
Like you said, it is sending a request to the mail server, waiting for a response, then sending another request and eventually timing out. The current mail code is somewhat old in terms of Laravel and doesn't support having more than one CC or BCC recipient, which you can see here (https://github.com/matrix-msu/Kora3/blob/28154a45f6562ba17bc8314ee43e6c174681e422/vendor/laravel/framework/src/Illuminate/Notifications/Messages/MailMessage.php#L173).

To have multiple BCC recipients, I implemented a slightly different and more recent Laravel mail function which is pretty neat because it simplifies/cleans up the mail code, has built-in queueing support (not being used right now), and uses Mailable (see app/Email) which introduces a more modular way to handle emails.
I'm not completely sure how this works under-the-hood, but I think it's only sending 1 request to the mail server. 
I think it's worth testing this on the live-site to see if this'll fix the 504.

Installation admin gets the email, and all the other admins are BCC recipients.
This is on the issue-#348 branch (pushed to GitHub)

**Comment by fritosxii** [comment]
A small temp fix has been implemented for now. Larger mail improvements may have to be implemented in the future

**Comment by austintruchan** [comment]
Issue still exists when attempting to request project permissions.

**Comment by Dev34** [comment]
Unassigning myself because I'm not scheduled to work in the coming weeks

**Comment by Dev34** [comment]
If this is still a timeout issue - maybe creating a queue and sending the emails individually (no BCC) would fix the problem? 
Also, I _think_ the mail server tries for a full 30 seconds on invalid emails. That might also be causing this issue.

**Comment by fritosxii** [comment]
We went down to a single email, because it's a little quicker than doing a bunch of individuals.

Maybe look into why invalid emails do that, and then maybe we can modify the mail system to send once, and move on without "validating" the emails it sends to?

**Comment by Dev34** [comment]
I can take a look and see if invalid emails might be the issue

**Comment by Dev34** [comment]
Did some testing on my dev and these are my results (using iggy mail server):
I timed how long it took for the requesting user to hear back from the web server for a few variations of recipients:
![singlemailserverrequestprofiling](https://user-images.githubusercontent.com/12753141/50792059-c0759500-1291-11e9-9810-3fa466f9f100.PNG)
Fairly linear curve so the mail call is fairly consistent in duration, even with higher numbers of BCC recipients. It also didn't seem to be impacted by invalid emails, so that's good. 

I noticed that the installation admin always gets the email first and consistently within ~10 seconds of the mail call being executed. This is why we've noticed the install admins always getting the email.
This leads me to believe that the web server timing out the client's HTTP request is causing the current mail call to halt. I suspect this is because Laravel kills the thread handling the user's request, which is also handling the mail call. I initially thought the mail call would be safe on the mail server, but now I'm thinking Laravel sends the requests one recipient at a time to the mail server.

Even if the ongoing mail call is not interrupted, the ones after that would never execute because it doesn't execute the next mail call until the previous one finishes. There would only be more than 1 mail call if the user requests more than 1 project, because each project gets its own mail call.

I haven't been able to verify Laravel's exact behavior with the mail call, so the above stuff is just my best guess.

If I'm right, then the problem lies with having a super long HTTP request that doesn't get answered until the mail calls finish. If this is indeed the cause of the issue, then having the web server do the mail calls in the background is likely the best solution because then the user's HTTP request can be answered right away. This doesn't let us alert the user about the status of the mail call, however. 
You can do the mail calls in the background by queueing them, which requires one of the methods of queueing be set-up. (https://laravel.com/docs/5.7/queues).

@fritosxii Should I keep looking into this?

**Comment by Dev34** [comment]
I have pushed a branch named ```project-permissions-emails-#348``` that immediately pushes all the project permissions emails to a database table that acts as a queue. I believe it will fix the issue occurring on the live site.

In order for it to work, you run Laravel's queue worker in project root dir. I saw we have Linux screen so I created a screen with ```screen -S queue-worker``` and then executed ```php artisan queue:work``` when attached to that screen. For reference, a command to re-attach to the screen is ```screen -DR "queue-worker"``` and a screen session can be destroyed with ```screen -X -S [screen#] quit```. The queue worker needs to persist (otherwise the emails just get pushed to the queue but not sent), so you'll want to make sure not to destroy the screen session when you close the terminal.

When you're attached to the queue worker screen, you'll be able to see Laravel processing the emails. This should be helpful in determining if everything is working.
